File: 백준/Silver/27973. 지연 평가/지연 평가.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        int Q = Integer.parseInt(br.readLine());
        long small = 1;
        long add = 0;
        long multi = 1;

        for (int i = 0; i < Q; i++) {
            st = new StringTokenizer(br.readLine());

            int cmd = Integer.parseInt(st.nextToken());
            int num;

            if (cmd == 3) {
                System.out.println(small*multi+add);
            } else if (cmd == 1) {
                num = Integer.parseInt(st.nextToken());
                add *= num;
                multi *= num;
            } else if (cmd == 0) {
                num = Integer.parseInt(st.nextToken());
                add += num;
            } else {
                num = Integer.parseInt(st.nextToken());
                small += num;
            }
        }
    }
}

File: 백준/Silver/27973. 지연 평가/지연 평가.py
# 지연 평가
import sys
input = sys.stdin.readline

Q = int(input())
small = 1
add = 0
multi = 1

for i in range(Q):
    cmd = list(map(int,input().split()))

    if cmd[0] == 3:
        print(small*multi+add)
    elif cmd[0] == 1:
        add *= cmd[1]
        multi *= cmd[1]
    elif cmd[0] == 0:
        add += cmd[1]
    else:
        small += cmd[1]
